snyk/cli
# feat: ci mode differentiating permanent errors

Temporary failures might be treated differently from permanent failures in a CI context.

A temporary network partition might be retried with an exponential backoff delay in a CI script.

Such a temporary failure might even be skipped in the CI workflow, with assurance that security testing will be covered by Snyk monitors and reporting.

This option gives more flexibility and self-determination in choosing security posture. In some cases landing a critical fix is more important than waiting for a network partition or temporary outage to resolve.

This CI mode may be opted into by setting an environment variable SNYK_CI=1 when running the CLI. CI mode causes the following errors to terminate with a different exit code:

- Authorization errors, including entitlement errors (HTTP 401 and 403), will exit code 77 (EX_NOPERM)
- Other non-recoverable errors will exit code 69 (EX_UNAVAILABLE). This includes all other HTTP 4xx errors from Snyk APIs.

Recoverable errors (network connection errors, HTTP 5xx) will exit code 2.

The difference between CI and non-CI mode, is that the above errors will normally exit code 2.

## Pull Request Submission

Please check the boxes once done.

The pull request must:

- **Reviewer Documentation**
- **User facing Documentation**
- **Testing**

## Pull Request Review

All pull requests must undergo a thorough review process before being merged. 
The review process of the code PR should include code review, testing, and any necessary feedback or revisions. 
Pull request reviews of functionality developed in other teams only review the given documentation and test reports. 

Manual testing will not be performed by the reviewing team, and is the responsibility of the author of the PR.

For Node projects: It’s important to make sure changes in `package.json` are also affecting `package-lock.json` correctly.

****************************If a dependency is not necessary, don’t add it.**************************** 

When adding a new package as a dependency, make sure that the change is absolutely necessary. We would like to refrain from adding new dependencies when possible.
Documentation PRs in gitbook are reviewed by Snyk's content team. They will also advise on the best phrasing and structuring if needed.

## Pull Request Approval

Once a pull request has been reviewed and all necessary revisions have been made, it is approved for merging into 
the main codebase. The merging of the code PR is performed by the code owners, the merging of the documentation PR 
by our content writers.


## What does this PR do?

See preamble text above the template for a detailed explanation.

## Where should the reviewer start?

This is an opt-in re-routing of exit codes, for reasons given above.

## How should this be manually tested?

- Build the CLI with `make clean build`
- Test normal operation, should exit 0:
  - `SNYK_CI=1 snyk test`
  - `SNYK_CI=1 snyk container test alpine`
  - `SNYK_CI=1 snyk code test`
  - `SNYK_CI=1 snyk iac test`
- Test network partitions by setting `SNYK_API=http://localhost`. With or without SNYK_CI=1, you should get exit code 2.
- Test service unavailable by running a local HTTP server which always responds 500 on localhost:9000. `SNYK_API=http://localhost:9000` should exit code 2, with or without SNYK_CI=1.
- Test "unauthorized" by setting `SNYK_TOKEN=nope` in the above commands. With CI mode, you should get exit code 77 (EX_NOPERM), not 2. You can also use a token that doesn't have necessary entitlement (like code). You should get a 69 (EX_UNAVAILABLE) or 77 (EX_NOPERM), not 2.
- Test client errors by running a local HTTP server which always responds 400, 404, 422, etc. on localhost:9000. `SNYK_API=http://localhost:9000` should exit code 69 (EX_UNAVAILABLE) with SNYK_CI=1, not 2.

Note that without SNYK_CI=1, the above test cases should all exit code 2.

## Any background context you want to provide?

See PR preamble above for context; this is about giving more security posture options to CI pipeline owners and maintainers. 

CI exit code remapping is not comprehensive for all subcommands that the CLI supports -- only `test` commands. That should not be a blocker; additional commands might be added to CI mode as future enhancement.

Still a draft as this needs tests. Probably something scripted with Docker containers due to the nature of the changes. 

## What are the relevant tickets?

## Screenshots

## Additional questions
